spark/src/main/resources/error/delta-error-classes.json

```diff
@@ -170,6 +170,13 @@
     ],
     "sqlState" : "42703"
   },
+  "DELTA_CANNOT_DROP_CHECK_CONSTRAINT_FEATURE" : {
+    "message" : [
+      "Cannot drop the CHECK constraints table feature.",
+      "The following constraints must be dropped first: <constraints>."
```

Comment:
I was thinking that this a only a writer feature which means that can be safely ignored by readers that do not understand it. In that sense, dropping the feature should be allowed even if there are constraints in the table. They just need to be ignored. However, we do no need add a guard (if it is not already there) to only check constraints when the feature is enabled/supported. That is for writers that understand the feature.

On the other hand, if we do want to force this behavior it means the feature cannot be removed until the history is truncated. History validation and check is only done at the moment for all reader+writer features but there are plans to allow it for selected writer features.

Reply:
To add to that, it is good to drop constraints before dropping the feature but we should not check the history if there are constraints, i.e. `actionUsesFeature.`

Reply:
Requiring users to drop all constraints first before allowing the feature to be dropped is intentional. Otherwise, some users might not realize that dropping the `checkConstraints` feature will drop all check constraints on their table, or they may not realize that there are check constraints defined on their table.

We do not require all constraints to be removed from the history (see the added test), only that the current state of the table does not contain any check constraints. I have implemented `actionUsesFeature`, but AFAIK it is never called right now, as `historyContainsFeature` is only called for reader features (and this is a writer feature).